src/backend/sgx/attestation.rs

```diff
@@ -1,16 +1,83 @@
 // SPDX-License-Identifier: Apache-2.0
 
+// Credit to: https://github.com/fortanix/rust-sgx/tree/master/aesm-client
+// for examples of AESM Requests.
+
+use crate::protobuf::aesm_proto::{
+    Request, Request_InitQuoteRequest, Response, Response_InitQuoteResponse,
+};
 use crate::syscall::{SGX_DUMMY_QUOTE, SGX_DUMMY_TI, SGX_QUOTE_SIZE, SGX_TI_SIZE};
 
+use std::io::{Error, ErrorKind, Read, Write};
+use std::mem::size_of;
+use std::os::unix::net::UnixStream;
 use std::slice::from_raw_parts_mut;
 
+use protobuf::Message;
+
+const AESM_SOCKET: &str = "/var/run/aesmd/aesm.socket";
+
 /// Fills the Target Info of the QE into the output buffer specified and
 /// returns the number of bytes written.
-// TODO: Replace synthetic TargetInfo with real TargetInfo.
-fn get_ti(out_buf: &mut [u8]) -> usize {
+fn get_ti(out_buf: &mut [u8]) -> Result<usize, Error> {
     assert_eq!(out_buf.len(), SGX_TI_SIZE, "Invalid size of output buffer");
-    out_buf.copy_from_slice(&SGX_DUMMY_TI);
-    SGX_TI_SIZE
+
+    // If unable to connect to the AESM daemon, return dummy value
+    let mut stream = match UnixStream::connect(AESM_SOCKET) {
+        Ok(s) => s,
+        Err(_) => {
+            out_buf.copy_from_slice(&SGX_DUMMY_TI);
+            return Ok(SGX_TI_SIZE);
+        }
+    };
+
+    // Set an Init Quote Request
+    let mut req = Request::new();
+    let mut msg = Request_InitQuoteRequest::new();
+    msg.set_timeout(1_000_000);
+    req.set_initQuoteReq(msg);
+
+    // Set up Writer
+    let mut buf_wrtr = vec![0u8; size_of::<u32>()];
+    match req.write_to_writer(&mut buf_wrtr) {
+        Ok(_) => {}
+        Err(e) => {
+            eprintln!("Error getting TargetInfo: {:#?}", e);
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                "Invalid Init Quote Request",
```

Comment:
This can be folded into the `Err` that's returned below with `format!`. I don't think we should have syscall implementations printing inside or outside of the keep unless that syscall is a `write` syscall.

Reply:
Fixed!